File: frontend/src/app/services/activity.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface CategoryStat {
  total: number;
  solved: number;
  points: number;
}

export interface SolveEntry {
  challenge_id: string;
  challenge_title: string;
  category: string;
  points: number;
  solved_at: string;
}

export interface Achievement {
  id: string;
  user_id: string;
  type: string;
  name: string;
  description: string;
  icon: string;
  earned_at: string;
}

export interface UserActivity {
  user_id: string;
  username: string;
  total_solves: number;
  total_points: number;
  category_progress: { [key: string]: CategoryStat };
  recent_solves: SolveEntry[];
  achievements: Achievement[];
  rank: number;
  solve_streak: number;
}

@Injectable({
  providedIn: 'root'
})
export class ActivityService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) { }

  getMyActivity(): Observable<UserActivity> {
    return this.http.get<UserActivity>(`${this.apiUrl}/activity/me`);
  }

  getMyAchievements(): Observable<Achievement[]> {
    return this.http.get<Achievement[]>(`${this.apiUrl}/achievements/me`);
  }
}

File: backend/internal/models/activity.go
package models

import "time"

type UserActivity struct {
	UserID           string                  `json:"user_id"`
	Username         string                  `json:"username"`
	TotalSolves      int                     `json:"total_solves"`
	TotalPoints      int                     `json:"total_points"`
	CategoryProgress map[string]CategoryStat `json:"category_progress"`
	RecentSolves     []SolveEntry            `json:"recent_solves"`
	Achievements     []Achievement           `json:"achievements"`
	Rank             int                     `json:"rank"`
	SolveStreak      int                     `json:"solve_streak"`
}

type CategoryStat struct {
	Total  int `json:"total"`
	Solved int `json:"solved"`
	Points int `json:"points"`
}

type SolveEntry struct {
	ChallengeID    string    `json:"challenge_id"`
	ChallengeTitle string    `json:"challenge_title"`
	Category       string    `json:"category"`
	Points         int       `json:"points"`
	SolvedAt       time.Time `json:"solved_at"`
}

type AdminAnalytics struct {
	TotalUsers          int                   `json:"total_users"`
	TotalTeams          int                   `json:"total_teams"`
	TotalChallenges     int                   `json:"total_challenges"`
	TotalSubmissions    int                   `json:"total_submissions"`
	TotalCorrect        int                   `json:"total_correct"`
	SuccessRate         float64               `json:"success_rate"`
	ChallengePopularity []ChallengePopularity `json:"challenge_popularity"`
	CategoryBreakdown   map[string]int        `json:"category_breakdown"`
	DifficultyBreakdown map[string]int        `json:"difficulty_breakdown"`
	RecentActivity      []RecentActivityEntry `json:"recent_activity"`
	SolvesOverTime      []TimeSeriesEntry     `json:"solves_over_time"`
	SubmissionsOverTime []TimeSeriesEntry     `json:"submissions_over_time"`
	ActiveUsers         int                   `json:"active_users"`
	BannedUsers         int                   `json:"banned_users"`
	VerifiedUsers       int                   `json:"verified_users"`
	AdminCount          int                   `json:"admin_count"`
	NewUsersToday       int                   `json:"new_users_today"`
	NewUsersThisWeek    int                   `json:"new_users_this_week"`
	NewTeamsToday       int                   `json:"new_teams_today"`
	NewTeamsThisWeek    int                   `json:"new_teams_this_week"`
	SubmissionsToday    int                   `json:"submissions_today"`
	SolvesToday         int                   `json:"solves_today"`
	AverageTeamSize     float64               `json:"average_team_size"`
	UserGrowth          []TimeSeriesEntry     `json:"user_growth"`
	TeamGrowth          []TimeSeriesEntry     `json:"team_growth"`
	TopTeams            []TeamStats           `json:"top_teams"`
	TopUsers            []UserStats           `json:"top_users"`
}

type ChallengePopularity struct {
	ChallengeID  string  `json:"challenge_id"`
	Title        string  `json:"title"`
	Category     string  `json:"category"`
	SolveCount   int     `json:"solve_count"`
	AttemptCount int     `json:"attempt_count"`
	SuccessRate  float64 `json:"success_rate"`
}

type RecentActivityEntry struct {
	UserID        string    `json:"user_id"`
	Username      string    `json:"username"`
	Action        string    `json:"action"`
	ChallengeID   string    `json:"challenge_id,omitempty"`
	ChallengeName string    `json:"challenge_name,omitempty"`
	Timestamp     time.Time `json:"timestamp"`
}

type TimeSeriesEntry struct {
	Date  string `json:"date"`
	Count int    `json:"count"`
}

type TeamStats struct {
	TeamID      string `json:"team_id"`
	Name        string `json:"name"`
	Score       int    `json:"score"`
	MemberCount int    `json:"member_count"`
	SolveCount  int    `json:"solve_count"`
}

type UserStats struct {
	UserID     string `json:"user_id"`
	Username   string `json:"username"`
	Score      int    `json:"score"`
	SolveCount int    `json:"solve_count"`
}
